Garrison rework

jaj22 opened the pull request:
### What type of PR is this.
1. [ ] Bug
2. [X] Change
3. [X] Enhancement

### What have you changed and why?
Complete rework of the underlying garrison logic. The general intention is to tighten up the accounting and control flow while increasing persistence. The initial functionality goal is to replicate all current features while adding general persistence for enemy garrison vehicles, persisting rebel vehicles places in any garrison, and improving performance through better pause/unpause and storing rebel statics in garrisons.

The basic structure is a split between server-side unspawned garrison data (stored in A3A_garrison, manipulated by unscheduled garrisonServer functions) and local spawned garrison data (stored in A3A_activeGarrison, maintained by scheduled garrisonLocal functions. Despite being scheduled, the garrisonLocal functions are serialized using a garrison op queue system. So for example, a function that pauses a garrison cannot be called until the spawn of the garrison is completed, even though the spawn may take several seconds.

Placeholder: I'll add some stuff about the data structures here later.

Code is currently "mostly working", as in it'd probably take me more than 10 seconds to find a new bug. Doesn't mean that it'd take anyone else that long. Usage patterns vary. Changes are massive. Save compatibility is fairly extensive despite major data differences, but backwards-only. You cannot save a game in this version and then use it in previous versions.

Known-missing/broken functionality:
- ~~Sniper positions (eg. the dudes on radio towers).~~
- ~~Specops UAVs.~~
- ~~Add mortar button on UIs will not add a mortar. This is intentional and the buttons need removing.~~
- ~~selectGarrisonVehicleType is hot garbage.~~
- ~~Runway spawns.~~
- ~~Need to handle case where faction lacks helis.~~
- ~~Allow rebel garrisons to use weapons on parked vehicles~~
- ~~Enemy garrison sizes are currently too low due to removal of freebies.~~
- ~~Reinf code currently has no vehicle upgrade capability or ability to reinforce minor sites.~~

Stretch goals:
- ~~Police stations.~~
- ~~Improvements to intra-garrison communications and mortar usage.~~
- ~~Civ car persistence.~~
- Partial spawn options (artillery, AA).

### Please specify which Issue this PR Resolves.
closes #XXXX

### Please verify the following and ensure all checks are completed.
1. [X] Have you loaded the mission in LAN host?
2. [X] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [X] No
2. [ ] Yes (Please provide further detail below.)

jaj22 committed:
Initial mostly-working rebel garrison rewrite

jaj22 committed:
Merge branch 'unstable' into new-garrisons-1

jaj22 committed:
Blind commit

jaj22 committed:
Merge branch 'unstable' into new-garrisons-1

jaj22 committed:
Post-merge fixups 1

jaj22 committed:
Remove minor sites from markersX

jaj22 committed:
File renaming, lots of fixup

jaj22 committed:
Second half updates

jaj22 committed:
Spawn/despawn/capture testing

jaj22 committed:
Merge branch 'unstable' into new-garrisons-1

jaj22 committed:
hqDialog add/remove, roadblock fixes

jaj22 committed:
distance & GC updates, rebel HQ perma-spawn

jaj22 committed:
Fix new-game startup

jaj22 committed:
Cleanup dead code, fix saving deleted garrisons and roadblock destruction logic

jaj22 committed:
Enemy group -> garrison and deep copy fixes

jaj22 committed:
Radio tower snipers & HC init order

jaj22 committed:
Re-add specops UAVs

jaj22 committed:
Remove mortar buttons in garrison menus

jaj22 committed:
Runway spawns and garrison missing vehicle type handling

jaj22 committed:
Runway spawns and better garrison vehicle selection

jaj22 committed:
Mostly working police stations and persistent civ cars

jaj22 committed:
Merge branch 'new-garrisons-1' of https://github.com/jaj22/A3-Antistasi into new-garrisons-1

jaj22 committed:
Increase garrison sizes, include medium groups, fix minor bugs

jaj22 committed:
Remove dead garrison spawning functions

jaj22 committed:
Merge statics/vehicles, implement rebel garrison vehicle usage

jaj22 committed:
Add garrison-managed responses, rebalance mortars

jaj22 committed:
Improved reinforcements and reorganized some related things

jaj22 committed:
Reinf improvements, add boat spawns

jaj22 committed:
AI mortar responses + reinf fixes

jaj22 committed:
Police station mission & destruction fixes

jaj22 committed:
Fix map info dialog

jaj22 committed:
Rework marker deletion, plus mortar, surrender fixes

jaj22 committed:
Merge branch 'unstable' into new-garrisons-1

jaj22 committed:
Replace automatic vehicle garrisoning with manual, slightly rough

jaj22 committed:
More new-garrison fixes:
- Make police station destruction trigger work more reliably.
- Reduce maximum spawn/unpause delay, especially at HQ.
- Workaround civ boat spawns not being deterministic.
- Fix logging errors in reinforceVehicle.
- Probably fix error in createAttackVehicle when the vehicle fails to move out of spawn.

jaj22 committed:
Fix global population reb/occ flip

jaj22 committed:
More new-garrison tweaks and fixes:
- Make building garrisons more inclined to remain in building.
- Nerf police station loot crates.
- Fix rebel garrison size not updating on kills.
- Fix statics not spawning in towers with intel.
- Remove surrender on cleanup. Now retreat-only.

jaj22 committed:
Patcom TARGET disable, capture rule tweaks and removal of 0/0 moveHQ refund message

jaj22 committed:
Malden and Tanoa police stations

jaj22 committed:
More new-garrison fixes:
- Remove utility items from garrison when garage-refunded.
- Prevent auto-garrison of drones.
- Set resource pool correctly for garrisoned vehicles.
- Make save-tagged utility items add to garrisons again.
- Fix findAttackTargets breaking when there's a rebel airfield with troops.
- Fix another createAttackVehicle error in waypoint-skip logic.

jaj22 committed:
Fix addGroup treating all reinforcements as 0 quality

jaj22 committed:
Fix minor compat issue with old saves

jaj22 committed:
Add Livonia police station buildings

jaj22 committed:
New-garrison fixes and police station tweaks:
- Fix bug where sites could flip to invaders when they weren't involved.
- Expand defender count radius a bit.
- Prevent garrison mortars firing on detection event.
- Fix bug if a police car place wasn't found in new-game init.
- Improve chance of finding a police station.

jaj22 committed:
Add police stations for original Chernarus maps

jaj22 committed:
Police stations working on Takistan, moved some towns

jaj22 committed:
Put a 6 cap on land vehicles on an airport for now

jaj22 committed:
Convert civ town vehicles when changing factions

jaj22 committed:
Sahrani police stations

jaj22 committed:
VT7 police stations

jaj22 committed:
Merge branch 'unstable' into new-garrisons-1

rautamiekka commented on A3A/addons/core/functions/Base/fn_actionGarrisonVehicle.sqf:
```diff
@@ -0,0 +1,12 @@
+
+[_marker, _vehicle] remoteExecCall ["A3A_fnc_garrisonServer_addVehicle", 2];
+
+[_titleStr, "Vehicle added to garrison. Unload any important cargo or inventory because this may not be preserved when you leave the area. Entering the vehicle will remove it from the garrison."] remoteExecCall ["A3A_fnc_customHint", _player];
```
Long strings like these are a prime example of what shoulda been shown through a locally-run function since the start instead.

jaj22 replied:
Yes, although in this case it's a local function and I don't know why I remoteExecCall'd the customHints. Too much sun probably.

rautamiekka commented on A3A/addons/core/functions/Base/fn_flagaction.sqf:
```diff
@@ -171,10 +171,14 @@ switch _typeX do
     case "static":
     {
-        private _cond = "(isPlayer _this) and (_target getVariable ['ownerSide', teamPlayer] == teamPlayer) and (isNull attachedTo _target) and ";
+        private _cond = "(isPlayer _this) and (_target getVariable ['ownerSide', teamPlayer] == teamPlayer) and (!isNil {_target getVariable 'markerX'}) and ";
```
You can use `isNil` like this [as of 'Arma 3' v2.18](https://community.bistudio.com/wiki/isNil#Syntax_3) (with v2.20 also allowing the `Control` & `Display` types with the same syntax) instead:
```suggestion
        private _cond = "(isPlayer _this) and (_target getVariable ['ownerSide', teamPlayer] == teamPlayer) and !(_target isNil 'markerX') and ";
```

jaj22 committed:
Anizay police stations

jaj22 committed:
Colombia and Kujari police stations

jaj22 committed:
Expand police station debug logging

jaj22 committed:
Tembelan police stations and airbase fix

jaj22 committed:
Fix VN inheritance issues in initStaticPlaces

Tiny-DM committed:
spe + vn police stations

jaj22 commented on A3A/addons/config_fixes/VN/config.cpp:
```diff
@@ -0,0 +1,51 @@
+        units[] = {};
+        weapons[] = {};
+        requiredVersion = REQUIRED_VERSION;
+        requiredAddons[] = {"vn_weapons", "cam_lao_nam_c"};
```
loadorder_f_vietnam

Tiny-DM committed:
loadorder + orientation fixes

jaj22 committed:
Napf police stations

jaj22 committed:
Merge branch 'new-garrisons-1' of https://github.com/jaj22/A3-Antistasi into new-garrisons-1

jaj22 committed:
Chernarus 2020 police stations

jaj22 committed:
Fix rebel retreat behaviour

jaj22 committed:
Pulau police stations and population adjustment

jaj22 committed:
Isla Duala police stations

jaj22 committed:
Regero police stations and popuation tweaks

jaj22 committed:
Weferlingen (& winter) police stations

jaj22 committed:
Fixes for issues found in new-garrison DS/HC testing:
- Fix A3A_spawnPlacesHM not being complete when sent to HC.
- Send A3A_garrisonSize to HC.
- Move singleAttack to server so that it doesn't break on findXXXAttackBases.
- Fix addToGarrison limit code not working on client.
- Fix troopsOnly flag not being passed correctly to garrisonLocal_clear.
- Fix garrisons vehicles being re-added to garrison when garrison crew are deleted.
- Hack-fix updateStatics getting called on non-yet-local units.

Other changes:
- Stringtable actionGarrisonVehicle.
- Use new isNil syntax to improve static flagactions.
- Speed up garrison op loop response time.
- Added unit locality debug logging.
- Reduce camp troop quality at lower war tiers.

jaj22 committed:
Fix petros moving around a bit on load

jaj22 committed:
Add proper headers to new-garrison

Tiny-DM reviewed (approved):
A full code review is impossible to do here; substantial solo playtesting ironed out a decent number of issues and it should be good for next stage of testing now. Approving.

jaj22 committed:
Add copyright notices to houses.hpp files
